packages/evm/contracts/Hashi.sol

```diff
@@ -34,61 +34,89 @@
 pragma solidity ^0.8.17;
 
 import { IOracleAdapter } from "./interfaces/IOracleAdapter.sol";
+import { IHashi } from "./interfaces/IHashi.sol";
 
-contract Hashi {
-    error NoOracleAdaptersGiven(address emitter);
-    error OracleDidNotReport(address emitter, IOracleAdapter oracleAdapter);
-    error OraclesDisagree(address emitter, IOracleAdapter oracleOne, IOracleAdapter oracleTwo);
-
-    /// @dev Returns the hash reported by a given oracle for a given ID.
-    /// @param oracleAdapter Address of the oracle adapter to query.
-    /// @param domain Id of the domain to query.
-    /// @param id ID for which to return a hash.
-    /// @return hash Hash reported by the given oracle adapter for the given ID number.
-    function getHashFromOracle(
-        IOracleAdapter oracleAdapter,
+contract Hashi is IHashi {
+    /// @inheritdoc IHashi
+    function checkHashWithThresholdFromOracles(
         uint256 domain,
-        uint256 id
-    ) public view returns (bytes32 hash) {
-        hash = oracleAdapter.getHashFromOracle(domain, id);
+        uint256 id,
+        uint256 threshold,
+        IOracleAdapter[] calldata oracleAdapters
+    ) external view returns (bool) {
+        if (oracleAdapters.length < threshold) return false;
+
+        bytes32[] memory hashes = new bytes32[](oracleAdapters.length);
+        for (uint256 i = 0; i < oracleAdapters.length; ) {
+            try oracleAdapters[i].getHashFromOracle(domain, id) returns (bytes32 hash) {
+                hashes[i] = hash;
+            } catch {}
+            unchecked {
+                ++i;
+            }
+        }
+
+        for (uint256 i = 0; i < hashes.length; ) {
+            bytes32 baseHash = hashes[i];
+            if (baseHash == bytes32(0)) {
+                unchecked {
+                    ++i;
+                }
+                continue;
+            }
+
+            uint256 num = 0;
+            for (uint256 j = 0; j < hashes.length; ) {
+                if (baseHash == hashes[j]) {
+                    unchecked {
+                        ++num;
+                    }
+                    if (num == threshold) return true;
```

Comment:
what if threshold = 0 && oracleAdapters.length > 0? This would fail, but is it the desired behaviour? In other terms, is threshold = 0 considered valid? I can see some use cases would it would. Replacing "==" here with ">" may be a good start, but not sufficient. A unit test covering these edge cases would be good

Reply:
the best approach here could be to make it clear in the code that threshold should always be >= 1